### ISSUE TITLE
Migrations up() do not respect "schema" annotation variable

### ISSUE THREAD

**Issue by guillaume-a** [doctrine/dbal#3482]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.9.2

#### Summary

I'm currently working on a project with multiple databases so I specifies schema in my @table.

#### Current behaviour

doctrine:migration:diff`generates a migration file, but up() does not use "schema"

#### How to reproduce

Create an entity like this.

> /**
>  * @ORM\Entity(repositoryClass="App\Repository\TinyElephantRepository")
>  * @ORM\Table(schema="sf")
>  */
> class TinyElephant

Generate a diff on an existing table
Output is (simplified): 

```
    public function up(Schema $schema) : void
    {
        $this->addSql('ALTER TABLE tiny_elephant ADD name VARCHAR(255) DEFAULT NULL');
    }

    public function down(Schema $schema) : void
    {
        $this->addSql('ALTER TABLE sf.tiny_elephant DROP name');
    }
```

#### Expected behaviour

Using schema in up() too.

```
    public function up(Schema $schema) : void
    {
        $this->addSql('ALTER TABLE sf.tiny_elephant ADD name VARCHAR(255) DEFAULT NULL');
    }

```

The issue is diff is made only from the schema and mysql do not supports "namespaces".
Regards.

**Comment by Ocramius** [comment]
@guillaume-a can you try and isolate this with just the `AbstractSchemaManager`?

**Comment by guillaume-a** [comment]
I'm sorry, how to do this exactly ?
Thx.

**Comment by Ocramius** [comment]
Ideally through a test scenario, but first you should look at where the culprit is (maybe with a debugger)

**Comment by guillaume-a** [comment]
Ok, i'll give a test with deep dive in the code :)

`AbstractSchemaManager::listTableDetails($tableName)`

returns a Table($tableName, ...)

this table does not have a _namespace because _setName($tableName) checks the '.' to separate namespace and name

`Table::__construct()` (call _setNam line 61)
`AbstractAsset::_setName()` (check '.' at line 51)

So let's back a little because AbstractSchemaManager seems to to the job.
The probleme is its caller.

`AbstractSchemaManager::listTableNames()`

```
returns [
  0 => 'migration_versions',
  1 => 'tiny_elephant',
]
```

Here there is no schema.
Because in mysql

`$sql = $this->_platform->getListTablesSQL();`

returns SHOW FULL TABLES WHERE Table_type = 'BASE TABLE' which lists tables without db name.

we have this name in : 

`(AbstractSchemaManager) $this->_conn->params['dbname']`

Maybe add a loop here to add dbname before table name ?

Regards.

**Comment by Ocramius** [comment]
Wait, so is `sf.` the schema you have selected while connecting? What if you use a different schema while connecting, and then generate a migration?

**Comment by guillaume-a** [comment]
Yes, but the schema is forced with @Table(schema="sf") in my entity.

Legacy project with multiple databases.

Le sam. 16 mars 2019 à 16:48, Marco Pivetta <notifications@github.com> a
écrit :
